# Titanic — exercícios avançados de merge em Pandas

Objetivo: transformar perguntas em etapas, identificar chaves, cruzar tabelas e conferir se a junção preservou os dados.

Execute a preparação, tente os seis exercícios e depois consulte o gabarito. As tabelas de passageiros, viagens e resultados são separadas a partir do **train.csv do Titanic** para simular fontes distintas. A tabela de portos é um cadastro didático criado neste notebook. Nenhuma fonte externa de passageiros é inventada.

**Antes de cada código:** escreva o que tem, o que quer, a chave de ligação, a cardinalidade esperada e os passos em português. Uma linha representa um passageiro? Um bilhete? Um grupo? Responda isso antes de contar.

## 1 Ler os arquivos locais

Extraia o ZIP inteiro e abra o notebook dentro da pasta extraída. A subpasta dados deve ficar ao lado dele. Este notebook não instala bibliotecas nem acessa o Kaggle. Use um kernel que já tenha pandas e IPython.

In [3]:
from pathlib import Path
import pandas as pd
from IPython.display import display


train = pd.read_csv("train.csv")
assert train["PassengerId"].is_unique
assert train["Survived"].isin([0, 1]).all()
print("Dimensões:", train.shape)
display(train.head())
display(train.isna().sum().rename("nulos").to_frame())

Dimensões: (891, 12)


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


,nulos
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


## 2 Preparar as tabelas dos exercícios

In [ ]:
passageiros = pd.read_csv(pasta / "passageiros.csv")
viagens = pd.read_csv(pasta / "viagens.csv")
resultados = pd.read_csv(pasta / "resultados.csv")
portos = pd.read_csv(pasta / "portos.csv")
for nome, tabela in {"passageiros": passageiros, "viagens": viagens,
                     "resultados": resultados, "portos": portos}.items():
    print(nome, tabela.shape)
    display(tabela.head())

passageiros (891, 6)


,PassengerId,Name,Sex,Age,SibSp,Parch
0,1,"Braund, Mr. Owen Harris",male,22.0,1,0
1,2,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0
2,3,"Heikkinen, Miss. Laina",female,26.0,0,0
3,4,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0
4,5,"Allen, Mr. William Henry",male,35.0,0,0


viagens (891, 5)


,PassengerId,Pclass,Ticket,Fare,Embarked
0,1,3,A/5 21171,7.2500,S
1,2,1,PC 17599,71.2833,C
2,3,3,STON/O2. 3101282,7.9250,S
3,4,1,113803,53.1000,S
4,5,3,373450,8.0500,S


resultados (891, 2)


,PassengerId,Survived
0,1,0
1,2,1
2,3,1
3,4,1
4,5,0


portos (3, 2)


,codigo_porto,nome_porto
0,C,Cherbourg
1,Q,Queenstown
2,S,Southampton


### Referência rápida
| Junção | O que mantém |
|---|---|
| left | Todas as linhas da esquerda e correspondências da direita |
| inner | Somente correspondências |
| outer | Chaves dos dois lados, mesmo sem correspondência |

`validate="one_to_one"` exige chaves únicas dos dois lados; `many_to_one` permite repetição na esquerda e exige unicidade na direita. `indicator=True` mostra a origem da correspondência. No Pandas, chaves nulas podem casar entre si; por isso o cadastro de portos não terá uma chave nula. Uma chave duplicada pode multiplicar linhas e distorcer somas.

## 3 Exercícios — tente antes de abrir o gabarito

### Exercício 1 Reconstruir a base analítica

Junte passageiros, viagens e resultados. Preserve todos os passageiros e confirme que há uma linha por PassengerId. Mostre nome, classe, porto e sobrevivência.

**Como pensar:** Que coluna existe nas três tabelas? Qual é a unidade de cada tabela? Como verificar linhas perdidas ou duplicadas?

Preencha: dados disponíveis → resultado esperado → chave → cardinalidade → passos → conferência.

In [ ]:
# Escreva sua solução aqui.


### Exercício 2 Enriquecer os portos sem perder passageiros

Adicione nome_porto à base. As chaves têm nomes diferentes. Conte os passageiros sem correspondência, mostre seus IDs e compare o número de linhas de left e inner. Não preencha valores ausentes com um porto escolhido arbitrariamente.

**Como pensar:** Qual tabela deve ficar à esquerda? Por que a chave se repete entre passageiros, mas não no cadastro?

Preencha: dados disponíveis → resultado esperado → chave → cardinalidade → passos → conferência.

In [ ]:
# Escreva sua solução aqui.


### Exercício 3 Bilhetes compartilhados

Crie uma tabela com uma linha por Ticket, contendo o número de passageiros daquele bilhete no train.csv. Faça merge desse resumo com a base, selecione bilhetes com pelo menos 3 passageiros e calcule sobreviventes e taxa de sobrevivência por bilhete.

**Como pensar:** Contar linhas do train representa todos os passageiros históricos? Depois do merge, em que coluna os valores se repetem?

Preencha: dados disponíveis → resultado esperado → chave → cardinalidade → passos → conferência.

In [ ]:
# Escreva sua solução aqui.


### Exercício 4 Comparar a tarifa com seu grupo

Calcule a mediana de Fare por Pclass e Sex, ignorando nulos. Junte o resumo à base usando as duas chaves e encontre passageiros cuja tarifa seja maior que a mediana do grupo. Mostre ID, classe, sexo, tarifa, mediana e diferença.

**Como pensar:** Por que Pclass sozinha não identifica cada linha do resumo? O que aconteceria se usássemos apenas essa chave?

Preencha: dados disponíveis → resultado esperado → chave → cardinalidade → passos → conferência.

In [ ]:
# Escreva sua solução aqui.


### Exercício 5 Auditar resultados ausentes e chaves órfãs

Crie uma cópia de resultados sem o primeiro passageiro e acrescente uma linha com um ID novo que não esteja na base. Faça outer merge e liste left_only e right_only. Demonstre por que ausência de resultado não significa Survived = 0.

**Como pensar:** O outer permite enxergar problemas dos dois lados? Quais duas anomalias foram introduzidas de propósito?

Preencha: dados disponíveis → resultado esperado → chave → cardinalidade → passos → conferência.

In [ ]:
# Escreva sua solução aqui.


### Exercício 6 Investigar um merge que infla contagens

Duplique a linha do primeiro porto no cadastro. Tente a junção com many_to_one e capture MergeError. Depois execute sem validação, meça a inflação de linhas e identifique os IDs repetidos. Corrija o cadastro e confira a base final.

**Como pensar:** Por que um cadastro com duas linhas para a mesma chave multiplica passageiros? Quando remover duplicatas seria inadequado?

Preencha: dados disponíveis → resultado esperado → chave → cardinalidade → passos → conferência.

In [ ]:
# Escreva sua solução aqui.


## 4 Gabarito comentado
Execute em ordem. Os exercícios 2 a 6 usam a base construída no exercício 1. Os erros intencionais são capturados para não interromper a execução.

### Solução 1 Reconstruir a base analítica

Passos: escolher PassengerId → juntar viagens → juntar resultados → comparar contagem e unicidade. Todas as tabelas foram separadas do mesmo train, então esperamos correspondência completa.

In [ ]:
base = passageiros.merge(viagens, on="PassengerId", how="left", validate="one_to_one")
base = base.merge(resultados, on="PassengerId", how="left", validate="one_to_one", indicator=True)
assert len(base) == len(passageiros)
assert base["PassengerId"].is_unique
assert base["_merge"].eq("both").all()
base = base.drop(columns="_merge")
display(base[["PassengerId", "Name", "Pclass", "Embarked", "Survived"]].head())

,PassengerId,Name,Pclass,Embarked,Survived
0,1,"Braund, Mr. Owen Harris",3,S,0
1,2,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,C,1
2,3,"Heikkinen, Miss. Laina",3,S,1
3,4,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,S,1
4,5,"Allen, Mr. William Henry",3,S,0


### Solução 2 Enriquecer os portos sem perder passageiros

Passos: manter passageiros à esquerda → ligar códigos com nomes diferentes → usar indicator → inspecionar ausências. O inner exclui quem não encontrou cadastro; isso muda a população analisada.

In [ ]:
base_portos = base.merge(portos, left_on="Embarked", right_on="codigo_porto",
                         how="left", validate="many_to_one", indicator=True)
assert len(base_portos) == len(base)
sem_porto = base_portos.loc[base_portos["_merge"].eq("left_only"), ["PassengerId", "Embarked"]]
display(sem_porto)
base_inner = base.merge(portos, left_on="Embarked", right_on="codigo_porto",
                        how="inner", validate="many_to_one")
print("left:", len(base_portos), "inner:", len(base_inner), "sem correspondência:", len(sem_porto))
assert len(base_portos) - len(base_inner) == len(sem_porto)

,PassengerId,Embarked
61,62,NaN
829,830,NaN


left: 891 inner: 889 sem correspondência: 2


### Solução 3 Bilhetes compartilhados

Passos: agrupar viagens por bilhete → contar passageiros → enriquecer cada passageiro → filtrar → reagrupar. A contagem vale somente para o train. Mesmo Ticket não comprova vínculo familiar. Não some passageiros_no_train após o merge: o total do grupo está repetido em cada passageiro.

In [ ]:
bilhetes = (viagens.groupby("Ticket", dropna=False)
            .agg(passageiros_no_train=("PassengerId", "size")).reset_index())
base_bilhetes = base.merge(bilhetes, on="Ticket", how="left", validate="many_to_one")
assert len(base_bilhetes) == len(base)
compartilhados = base_bilhetes.loc[base_bilhetes["passageiros_no_train"] >= 3]
resumo_bilhetes = (compartilhados.groupby("Ticket", dropna=False)
    .agg(passageiros=("PassengerId", "size"), sobreviventes=("Survived", "sum"),
         taxa_sobrevivencia=("Survived", "mean")).reset_index())
resumo_bilhetes["taxa_percentual"] = resumo_bilhetes["taxa_sobrevivencia"] * 100
display(resumo_bilhetes.sort_values(["passageiros", "Ticket"], ascending=[False, True]).head(10))
assert (resumo_bilhetes["sobreviventes"] <= resumo_bilhetes["passageiros"]).all()

,Ticket,passageiros,sobreviventes,taxa_sobrevivencia,taxa_percentual
5,1601,7,5,0.714286,71.428571
17,347082,7,0,0.000000,0.000000
29,CA. 2343,7,0,0.000000,0.000000
14,3101295,6,0,0.000000,0.000000
18,347088,6,0,0.000000,0.000000
28,CA 2144,6,0,0.000000,0.000000
24,382652,5,0,0.000000,0.000000
37,S.O.C. 14879,5,0,0.000000,0.000000
2,113760,4,4,1.000000,100.000000
3,113781,4,2,0.500000,50.000000


### Solução 4 Comparar a tarifa com seu grupo

Passos: agrupar por duas colunas → calcular mediana → fazer merge pela chave composta → calcular diferença → filtrar. Fare é a tarifa registrada no dataset; não inferimos preço individual de um bilhete compartilhado. Uma comparação com NaN não seleciona a linha; a mediana ignora valores ausentes.

In [ ]:
medianas = (base.groupby(["Pclass", "Sex"], dropna=False)
            .agg(mediana_tarifa=("Fare", "median")).reset_index())
comparacao = base.merge(medianas, on=["Pclass", "Sex"], how="left", validate="many_to_one")
assert len(comparacao) == len(base)
comparacao["diferenca"] = comparacao["Fare"] - comparacao["mediana_tarifa"]
acima = comparacao.loc[comparacao["Fare"] > comparacao["mediana_tarifa"],
                      ["PassengerId", "Pclass", "Sex", "Fare", "mediana_tarifa", "diferenca"]]
display(acima.sort_values("diferenca", ascending=False).head(10))
assert acima["diferenca"].gt(0).all()

,PassengerId,Pclass,Sex,Fare,mediana_tarifa,diferenca
679,680,1,male,512.3292,41.26250,471.06670
737,738,1,male,512.3292,41.26250,471.06670
258,259,1,female,512.3292,82.66455,429.66465
27,28,1,male,263.0000,41.26250,221.73750
438,439,1,male,263.0000,41.26250,221.73750
118,119,1,male,247.5208,41.26250,206.25830
557,558,1,male,227.5250,41.26250,186.26250
527,528,1,male,221.7792,41.26250,180.51670
88,89,1,female,263.0000,82.66455,180.33545
341,342,1,female,263.0000,82.66455,180.33545


### Solução 5 Auditar resultados ausentes e chaves órfãs

Passos: introduzir dois problemas controlados → juntar os dois lados → examinar o indicador. NaN representa resultado desconhecido neste cruzamento; preencher com zero mudaria o significado para morte observada.

In [ ]:
id_removido = resultados.iloc[0]["PassengerId"]
id_orfao = int(base["PassengerId"].max()) + 1
resultados_auditoria = pd.concat([
    resultados.loc[resultados["PassengerId"] != id_removido],
    pd.DataFrame({"PassengerId": [id_orfao], "Survived": [1]})
], ignore_index=True)
auditoria = passageiros.merge(resultados_auditoria, on="PassengerId", how="outer",
                               validate="one_to_one", indicator=True)
display(auditoria.loc[auditoria["_merge"] != "both", ["PassengerId", "Name", "Survived", "_merge"]])
assert auditoria["_merge"].eq("left_only").sum() == 1
assert auditoria["_merge"].eq("right_only").sum() == 1
assert auditoria.loc[auditoria["_merge"].eq("left_only"), "Survived"].isna().all()

,PassengerId,Name,Survived,_merge
0,1,"Braund, Mr. Owen Harris",NaN,left_only
891,892,NaN,1.0,right_only


### Solução 6 Investigar um merge que infla contagens

Passos: testar cardinalidade → observar multiplicação → identificar IDs repetidos → corrigir a origem. Se duas linhas tiverem o mesmo código e nomes conflitantes, drop_duplicates não resolve: é necessário decidir qual registro é válido.

In [ ]:
portos_duplicados = pd.concat([portos, portos.iloc[[0]]], ignore_index=True)
try:
    base.merge(portos_duplicados, left_on="Embarked", right_on="codigo_porto",
               how="left", validate="many_to_one")
except pd.errors.MergeError as erro:
    print("Erro esperado de cardinalidade:", erro)
else:
    raise AssertionError("A validação deveria detectar a chave duplicada.")
inflada = base.merge(portos_duplicados, left_on="Embarked", right_on="codigo_porto", how="left")
print("Linhas originais:", len(base), "após merge:", len(inflada), "inflação:", len(inflada)-len(base))
display(inflada.loc[inflada["PassengerId"].duplicated(keep=False), ["PassengerId", "Embarked", "nome_porto"]].head(10))
# Neste caso a duplicação é exata e foi criada artificialmente.
portos_corrigidos = portos_duplicados.drop_duplicates()
assert portos_corrigidos["codigo_porto"].is_unique
corrigida = base.merge(portos_corrigidos, left_on="Embarked", right_on="codigo_porto",
                       how="left", validate="many_to_one")
assert len(corrigida) == len(base)
assert corrigida["PassengerId"].is_unique
assert len(inflada)-len(base) == base["Embarked"].eq(portos.iloc[0]["codigo_porto"]).sum()

Erro esperado de cardinalidade: Merge keys are not unique in right dataset; not a many-to-one merge
Linhas originais: 891 após merge: 1059 inflação: 168


,PassengerId,Embarked,nome_porto
1,2,C,Cherbourg
2,2,C,Cherbourg
10,10,C,Cherbourg
11,10,C,Cherbourg
21,20,C,Cherbourg
22,20,C,Cherbourg
29,27,C,Cherbourg
30,27,C,Cherbourg
34,31,C,Cherbourg
35,31,C,Cherbourg


## 5 Desafio final
Construa um relatório por nome_porto e Pclass com passageiros, sobreviventes, taxa de sobrevivência e mediana de idade. Preserve passageiros sem porto e mostre-os como "Não informado" somente no rótulo do relatório. Inclua o número de idades conhecidas em cada grupo e explique por que isso importa para a mediana.

Antes de agrupar, verifique uma linha por passageiro. Na comparação entre grupos, apresente também o tamanho da amostra; os resultados descrevem associações no train e não estabelecem causalidade.

In [ ]:
# Sua solução do desafio final.


### Gabarito do desafio final

In [ ]:
relatorio_base = base.merge(portos, left_on="Embarked", right_on="codigo_porto",
                           how="left", validate="many_to_one")
assert relatorio_base["PassengerId"].is_unique
relatorio_base["nome_porto"] = relatorio_base["nome_porto"].fillna("Não informado")
relatorio = (relatorio_base.groupby(["nome_porto", "Pclass"], dropna=False)
    .agg(passageiros=("PassengerId", "size"), sobreviventes=("Survived", "sum"),
         taxa_sobrevivencia=("Survived", "mean"), mediana_idade=("Age", "median"),
         idades_conhecidas=("Age", "count")).reset_index())
relatorio["taxa_percentual"] = relatorio["taxa_sobrevivencia"] * 100
assert relatorio["passageiros"].sum() == len(base)
display(relatorio)
print("A mediana considera somente idades conhecidas; count exclui nulos e size conta linhas.")

,nome_porto,Pclass,passageiros,sobreviventes,taxa_sobrevivencia,mediana_idade,idades_conhecidas,taxa_percentual
0,Cherbourg,1,85,59,0.694118,36.5,74,69.411765
1,Cherbourg,2,17,9,0.529412,25.0,15,52.941176
2,Cherbourg,3,66,25,0.378788,20.0,41,37.878788
3,Não informado,1,2,2,1.000000,50.0,2,100.000000
4,Queenstown,1,2,1,0.500000,38.5,2,50.000000
5,Queenstown,2,3,2,0.666667,43.5,2,66.666667
6,Queenstown,3,72,27,0.375000,21.5,24,37.500000
7,Southampton,1,127,74,0.582677,37.0,108,58.267717
8,Southampton,2,164,76,0.463415,30.0,156,46.341463
9,Southampton,3,353,67,0.189802,25.0,290,18.980170


A mediana considera somente idades conhecidas; count exclui nulos e size conta linhas.


## Origem e limites

train.csv é uma cópia pública do dataset Titanic de 891 passageiros, obtida de https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv . Passageiros, viagens e resultados foram separados desse arquivo para a atividade. Portos é um cadastro didático. O notebook usa somente arquivos locais. As soluções foram executadas com os 891 registros incluídos. Contagens de bilhetes descrevem apenas essa amostra.

Documentação: https://pandas.pydata.org/docs/reference/api/pandas.merge.html